### DISCUSSION TITLE
Download Scripts

### DISCUSSION THREAD

**Discussion by pohzim** [WAM2layers/WAM2layers#337] [Q&A] [answered]
Hello all! My name is Luís. Hope you are all fine. First of all I'd like to say I never used github before, and came here because I am using WAM2layers and ran into some troubles I'd like to share with you. 

The first problem is related to the download scripts. The one provided here works really fine but takes too long because it requests one file per time. The ECMWF guidelines (lost the link, but will try to find) recommends download of whole months data in one file for efficiency. In my user experience the retrieval of one month of data with the script here provided (a file per day and variable) took almost a week and half, while retrieving whole months of data at once took approximately 3-4 hours. However, downloading data this way require splitting it, which I did with a python script.

The second problem I ran refers to the model level data. Longitude is retrieved in 0-360º format for model levels (figure below). 

![image](https://github.com/WAM2layers/WAM2layers/assets/70667042/db738366-0056-444d-bb6f-5e9588fbb640)

When trying to preprocess era5 raw data with the preprocess function, it returns the error below.

```
(amb1) poh@poh:~/era5_data$ wam2layers preprocess era5 config_file.yaml 
Welcome to WAM2layers.
Starting preprocessing ERA5 data.
2023-01-01 00:00:00
Traceback (most recent call last):
  File "/home/poh/miniconda3/envs/amb1/bin/wam2layers", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/c/Users/po/WAM2layers/src/wam2layers/cli.py", line 130, in era5
    prep_experiment(config_file)
  File "/mnt/c/Users/po/WAM2layers/src/wam2layers/preprocessing/era5.py", line 261, in prep_experiment
    cwv = q * dp / g  # (kg/m2)
          ~~^~~~
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/core/_typed_ops.py", line 251, in __mul__
    return self._binary_op(other, operator.mul)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/core/dataarray.py", line 4691, in _binary_op
    f(self.variable, other_variable_or_arraylike)
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/core/_typed_ops.py", line 487, in __mul__
    return self._binary_op(other, operator.mul)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/core/variable.py", line 2413, in _binary_op
    result = Variable(dims, new_data, attrs=attrs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/core/variable.py", line 365, in __init__
    super().__init__(
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/namedarray/core.py", line 253, in __init__
    self._dims = self._parse_dimensions(dims)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/poh/miniconda3/envs/amb1/lib/python3.11/site-packages/xarray/namedarray/core.py", line 481, in _parse_dimensions
    raise ValueError(
ValueError: dimensions ('level', 'latitude', 'longitude') must have the same length as the number of data dimensions, ndim=4
(amb1) poh@poh:~/era5_data$ 
```

In order to preprocess, I translated the data using and it preprocessed fine:

```
dataset = xarray.open_dataset('path.nc')
dataset = dataset.assign_coords(longitude=(((dataset.longitude + 180) % 360) -180))\
dataset = dataset.sortby(dataset.longitude)
```

So, about the question: Can I share the python script I used to download the data, translate and split? If yes, how can I do that? As said before, I never used github nor wrote codes and need some help with this.

Thanks for all!

**Comment by BSchilperoort** [comment]
Hi Luís, welcome!

Where did you retrieve the data from? If you download ERA5 data from the CDS (i.e. the provided script), the longitude should range from -180 to 180 degrees.

> The one provided here works really fine but takes too long because it requests one file per time. The ECMWF guidelines (lost the link, but will try to find) recommends download of whole months data in one file for efficiency. In my user experience the retrieval of one month of data with the script here provided (a file per day and variable) took almost a week and half

Only recently this has become an issue. The CDS is being migrated to new infrastructure, and small requests which would be finished in a minute can now take hours... Of course parallel requests can speed this up, but this was not a necessity before.

See [the message from ECMWF](https://forum.ecmwf.int/t/a-new-cds-soon-to-be-launched-expect-some-disruptions/1607) for more info. The queue time performance is available [here](https://eqc.climate.copernicus.eu/monitoring#highlevel_view;month-6), it's become atrocious.

**Comment by ruudvdent** [comment]
Dear Luís, welcome indeed!

To get back to your question:

> So, about the question: Can I share the python script I used to download the data, translate and split? If yes, how can I do that? As said before, I never used github nor wrote codes and need some help with this.

If the download performance by month is indeed much better, this would be very useful to other users. Please create a new branch, add the relevant scripts in the 'scripts' folder and create a pull request. This will be reviewed by one of us and then merged into the main branch. Requires a bit of Git knowledge, but I would recommend using VS code, which makes this all relatively easy. If all fails, you can also simply drop me an email.

**Reply by P-Myers** [reply]
Dear Ruud, Luis,

Has the script for a faster download process of the data been uploaded? I have not been able to locate it until now. I am working through a Google Cloud virtual machine, and the cost of downloading data this way over a long period is too expensive. Any guidance here would be greatly appreciated! 

Thank you!

**Reply by ruudvdent** [reply]
Luis has send me the files by email, but I haven't had time to look at them in detail or merge into the repository itself. I attach them here, but credits to Luis @pohzim. 
[script_download_ERA5_single_levels.txt](https://github.com/user-attachments/files/16638042/script_download_ERA5_single_levels.txt)

Note that @BSchilperoort has also started to look at another alternative in #401 , which may be useful as well.

**Reply by P-Myers** [reply]
Hi,

Thank you for sharing the script. I am having issues however only with downloading my model levels. I have successfully downloaded monthly datasets for the model levels, however they contain data for only two days (the first and last days of the month), and therefore are unusable. 

The download script I am using is very similar to the one shared in the 'scripts' folder, besides the fact that it downloads at a monthly resolution. I have ruled out the CDS switch to the new Beta platform as the issue, given that I have data downloaded from the previous platform that shares this trait.

Is this perhaps typical for the model level data when downloaded per month? 

As I search for a solution, I have begun downloading daily files for the model levels, which is unfortunately, very slow. 

Thank you,
Parker

**Reply by ruudvdent** [reply]
are you still having this particular issue?

**Comment by hidabril** [comment]
It works for me, thank you so much

**Comment by ruudvdent** [comment, marked as answer]
New CDS issue is now tackled in https://github.com/WAM2layers/WAM2layers/pull/429 and the ARCO ERA5 in https://github.com/WAM2layers/WAM2layers/pull/401 and https://github.com/WAM2layers/WAM2layers/pull/428. Hope this solves the problems. These changes will be incorporated when v4.2 is released. Thanks for sharing your experiences!
